keras/layers/recurrent.py

```diff
@@ -228,50 +233,61 @@ def __call__(self, inputs, initial_state=None, **kwargs):
         # then add it to the inputs and temporarily
         # modify the input spec to include the state.
         if initial_state is not None:
-            if hasattr(initial_state, '_keras_history'):
-                # Compute the full input spec, including state
-                input_spec = self.input_spec
-                state_spec = self.state_spec
-                if not isinstance(state_spec, list):
-                    state_spec = [state_spec]
-                self.input_spec = [input_spec] + state_spec
-
-                # Compute the full inputs, including state
-                if not isinstance(initial_state, (list, tuple)):
-                    initial_state = [initial_state]
-                inputs = [inputs] + list(initial_state)
-
-                # Perform the call
-                output = super(Recurrent, self).__call__(inputs, **kwargs)
-
-                # Restore original input spec
-                self.input_spec = input_spec
-                return output
-            else:
-                kwargs['initial_state'] = initial_state
+
+            # We need to build the layer so that state_spec exists.
+            with K.name_scope(self.name):
+                if not self.built:
+                    self.assert_input_compatibility(inputs)
+                    if hasattr(inputs, '_keras_shape'):
+                        input_shape = inputs._keras_shape
+                    elif hasattr(K, 'int_shape'):
+                        input_shape = K.int_shape(inputs)
+                    else:
+                        raise ValueError('You tried to call layer "' + self.name +
+                                         '". This layer has no information'
+                                         ' about its expected input shape, '
+                                         'and thus cannot be built. '
+                                         'You can build it manually via: '
+                                         '`layer.build(batch_input_shape)`')
+                    self.build(input_shape)
+                    self.built = True
+
+            # Compute the full input spec, including state
+            input_spec = self.input_spec
+            state_spec = self.state_spec
+            if not isinstance(state_spec, list):
+                state_spec = [state_spec]
+            self.input_spec = [input_spec] + state_spec
+
+            # Compute the full inputs, including state
+            if not isinstance(initial_state, (list, tuple)):
+                initial_state = [initial_state]
+            inputs = [inputs] + list(initial_state)
```

Comment:
We're not making an exception, RNN layers will behave like every other layer. We're just building an API. The API is that `RNN(inputs, initial_state=x)` should work as a way to set initial state tensors, independently of the value of `x` (Keras tensor or not). It's actually very simple to set up, via a switch between inputs and layer keyword arguments.